wire-ios-request-strategy/Sources/Request Strategies/Conversation/MLSConversationParticipantsService.swift

```diff
@@ -0,0 +1,142 @@
+////
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+
+// sourcery: AutoMockable
+protocol MLSConversationParticipantsServiceInterface {
+
+    func addParticipants(
+        _ users: [ZMUser],
+        to conversation: ZMConversation
+    ) async throws
+
+    func removeParticipant(
+        _ user: ZMUser,
+        from conversation: ZMConversation
+    ) async throws
+
+}
+
+enum MLSConversationParticipantsError: Error {
+    case ignoredUsers(users: Set<ZMUser>)
+    case invalidOperation
+}
+
+struct MLSConversationParticipantsService: MLSConversationParticipantsServiceInterface {
+
+    // MARK: - Properties
+
+    private let context: NSManagedObjectContext
+    private let clientIDsProvider: MLSClientIDsProviding
+    private let mlsService: MLSServiceInterface
+
+    // MARK: - Life cycle
+
+    init?(context: NSManagedObjectContext) {
+
+        guard let syncContext = context.performAndWait({ context.zm_sync }) else {
+            return nil
+        }
+
+        guard let mlsService = syncContext.performAndWait({ syncContext.mlsService }) else {
+            return nil
+        }
```

Comment:
To avoid doing this we could enforce the context to be the sync context

Reply:
We could also just inject the mls service and a context, and make it the responsibility of the caller to ensure it's the sync context.

Reply:
I agree. But it would just move the problem to `ConversationParticipantsService` and I don't see much value in that. I'd rather wait until `mlsService` is decoupled from the context